<a href="https://colab.research.google.com/github/ddy623/Kaggle-Projects/blob/main/titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder


In [ ]:
# Load Data
fpath = '/content/drive/MyDrive/Colab Notebooks/Data Science Project/train.csv'
df_train = pd.read_csv(fpath)
df_train.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [ ]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.45,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,Q


In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df_train.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Remove cabin from dataset

In [ ]:
df_train.drop('Cabin', axis=1, inplace=True)
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


### Handling Missing Values in 'Age' Column

First, let's examine the extent of missing data in the 'Age' column.

In [ ]:
print(f"Missing values in 'Age' column before imputation: {df['Age'].isnull().sum()}")

Missing values in 'Age' column before imputation: 0


Now, we will impute the missing 'Age' values using the median of the column. The median is often preferred over the mean for age, as it's less sensitive to outliers.

In [ ]:
median_age = df_train['Age'].median()
df_train['Age'].fillna(median_age, inplace=True)
print(f"Median Age used for imputation: {median_age}")

Median Age used for imputation: 28.0


/tmp/ipykernel_4700/1613488050.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_train['Age'].fillna(median_age, inplace=True)


Finally, let's verify that all missing values in the 'Age' column have been handled.

In [ ]:
print(f"Missing values in 'Age' column after imputation: {df_train['Age'].isnull().sum()}")
df_train.info()

Missing values in 'Age' column after imputation: 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [ ]:
# Check for missing values in 'Fare'
print(f"Missing values in 'Fare' column before imputation: {df_train['Fare'].isnull().sum()}")

# Impute missing 'Fare' values with the median
median_fare = df_train['Fare'].median()
df_train['Fare'].fillna(median_fare, inplace=True)
print(f"Median Fare used for imputation: {median_fare}")

# Verify that missing values have been handled
print(f"Missing values in 'Fare' column after imputation: {df_train['Fare'].isnull().sum()}")
df_train.info()

Missing values in 'Fare' column before imputation: 0
Median Fare used for imputation: 14.4542
Missing values in 'Fare' column after imputation: 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


/tmp/ipykernel_4700/2504306872.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_train['Fare'].fillna(median_fare, inplace=True)


In [ ]:
# Check for missing values in 'Embarked'
print(f"Missing values in 'Embarked' column before imputation: {df_train['Embarked'].isnull().sum()}")

# Impute missing 'Embarked' values with the mode
mode_embarked = df_train['Embarked'].mode()[0]
df_train['Embarked'].fillna(mode_embarked, inplace=True)
print(f"Mode Embarked used for imputation: {mode_embarked}")

Missing values in 'Embarked' column before imputation: 2
Mode Embarked used for imputation: S


/tmp/ipykernel_4700/3058260292.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_train['Embarked'].fillna(mode_embarked, inplace=True)


In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


*Upload* test data

In [ ]:
#Load test data
fpath = '/content/drive/MyDrive/Colab Notebooks/Data Science Project/test.csv'
df_test = pd.read_csv(fpath)
df_test.head()

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


In [ ]:
df_test.drop('Cabin', axis=1, inplace=True)
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(4)
memory usage: 32.8+ KB


In [ ]:
#Check for missing value in 'Fare'
print(f"Missing values in 'Fare' column before imputation: {df_test['Fare'].isnull().sum()}")

#Impute missing 'Fare' values with the median
median_fare = df_test['Fare'].median()
df_test['Fare'].fillna(median_fare, inplace=True)
print(f"Median Fare used for imputation: {median_fare}")

Missing values in 'Fare' column before imputation: 1
Median Fare used for imputation: 14.4542


/tmp/ipykernel_4700/1212946017.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_test['Fare'].fillna(median_fare, inplace=True)


In [ ]:
# Check for missing values in 'Age'
print(f"Missing values in 'Age' column before imputation: {df_test['Age'].isnull().sum()}")

# Impute missing 'Age' values with the median
median_age = df_test['Age'].median()
df_test['Age'].fillna(median_age, inplace=True)
print(f"Median Age used for imputation: {median_age}")

# Verify that missing values have been handled
print(f"Missing values in 'Age' column after imputation: {df_test['Age'].isnull().sum()}")
df_test.info()

Missing values in 'Age' column before imputation: 86
Median Age used for imputation: 27.0
Missing values in 'Age' column after imputation: 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          418 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         418 non-null    float64
 9   Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(4)
memory usage: 32.8+ KB


/tmp/ipykernel_4700/1039608418.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_test['Age'].fillna(median_age, inplace=True)


In [ ]:
from sklearn.preprocessing import LabelEncoder
import numpy as np

# --- Preprocessing for both df_train and df_test ---

# Combine for consistent preprocessing, especially for one-hot encoding
combined_df = pd.concat([df_train.drop('Survived', axis=1), df_test], ignore_index=True)

# Encode 'Sex' column
le = LabelEncoder()
combined_df['Sex'] = le.fit_transform(combined_df['Sex'])

# One-hot encode 'Embarked' column
combined_df = pd.get_dummies(combined_df, columns=['Embarked'], drop_first=True)

# Drop 'Name' and 'Ticket' as they are not used as features in this model
combined_df.drop(['Name', 'Ticket'], axis=1, inplace=True)

# Separate back into df_train_processed and df_test_processed
df_train_processed = combined_df.iloc[:len(df_train)]
df_test_processed = combined_df.iloc[len(df_train):]

# Align columns (this is crucial if one-hot encoding results in different columns due to unique values)
# For example, if Embarked_Q was only in train but not test
# However, for this dataset, it's generally fine as Q, S, C are in both.
# Still, it's good practice to ensure alignment if using a combined df approach for encoding.

# Features and Target for training
X_train = df_train_processed.drop('PassengerId', axis=1)
y_train = df_train['Survived']

# Features for testing (drop PassengerId as it's not a feature)
X_test = df_test_processed.drop('PassengerId', axis=1)

# --- Model Training ---
# Initialize and train a RandomForestClassifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# --- Prediction ---
y_pred = model.predict(X_test)

# --- Create Submission File ---
# Create a DataFrame for submission
submission_df = pd.DataFrame({'PassengerId': df_test['PassengerId'], 'Survived': y_pred})

# Ensure 'Survived' column is integer type as required by Kaggle
submission_df['Survived'] = submission_df['Survived'].astype(int)

# Save the submission file to CSV
submission_filename = 'submission.csv'
submission_df.to_csv(submission_filename, index=False)

print(f"Submission file '{submission_filename}' created successfully!")
print("First 5 rows of the submission file:")
print(submission_df.head())

# Download the file
files.download(submission_filename)

Submission file 'submission.csv' created successfully!
First 5 rows of the submission file:
   PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         1
4          896         0


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

How many people survived and how many died in the Titanic?

In [ ]:
print("In the 'Survived' column:")
print("0 = Did not survive (Death)")
print("1 = Survived")

In the 'Survived' column:
0 = Did not survive (Death)
1 = Survived


In [ ]:
num_survived = df_train['Survived'].sum()
print(f"Number of people who survived the Titanic: {int(num_survived)}")

Number of people who survived the Titanic: 342


In [ ]:
num_died = (df_train['Survived'] == 0).sum()
print(f"Number of people who died in the Titanic: {int(num_died)}")

Number of people who died in the Titanic: 549


Based on the data, 549 people died in the Titanic and 342 people survived.  